File: multi_agent/projects/im-desktop/src/services/messageForwardService.ts
/**
 * 消息合并转发服务
 * 支持多条消息合并转发、选择转发、逐条转发
 */

import { apiClient } from './apiClient';

export interface ForwardBundle {
  id: number;
  bundleId: string;
  sourceConversationId: number;
  targetConversationId?: number;
  createdBy: number;
  forwardType: 'MERGE' | 'SELECT' | 'INDIVIDUAL';
  title?: string;
  messageIds: number[];
  messageCount: number;
  hasMedia: boolean;
  mediaCount: number;
  status: 'DRAFT' | 'PENDING' | 'SENT' | 'FAILED' | 'CANCELLED';
  sendMode: 'MERGE' | 'SEPARATE';
  includeSenderInfo: boolean;
  includeTimestamp: boolean;
  anonymizeSenders: boolean;
  customComment?: string;
  createdAt: string;
  updatedAt: string;
  forwardedAt?: string;
}

export interface ForwardConfig {
  sendMode?: 'MERGE' | 'SEPARATE';
  includeSenderInfo?: boolean;
  includeTimestamp?: boolean;
  anonymizeSenders?: boolean;
  customComment?: string;
  title?: string;
}

export interface ForwardStats {
  total: number;
  successful: number;
  failed: number;
  byType: {
    MERGE: number;
    SELECT: number;
    INDIVIDUAL: number;
  };
}

export class MessageForwardService {
  private static instance: MessageForwardService;
  
  public static getInstance(): MessageForwardService {
    if (!MessageForwardService.instance) {
      MessageForwardService.instance = new MessageForwardService();
    }
    return MessageForwardService.instance;
  }
  
  /**
   * 创建转发草稿
   */
  async createDraft(
    sourceConversationId: number,
    forwardType: 'MERGE' | 'SELECT' | 'INDIVIDUAL',
    messageIds: number[]
  ): Promise<{
    success: boolean;
    bundleId: string;
    id: number;
    messageCount: number;
    createdAt: string;
  }> {
    try {
      const response = await apiClient.post('/message/forward/draft', messageIds, {
        params: { sourceConversationId, forwardType }
      });
      return response.data;
    } catch (error) {
      console.error('创建转发草稿失败:', error);
      throw new Error('创建转发草稿失败');
    }
  }
  
  /**
   * 添加消息到草稿
   */
  async addMessage(bundleId: string, messageId: number): Promise<{
    success: boolean;
    bundleId: string;
    messageCount: number;
    updatedAt: string;
  }> {
    try {
      const response = await apiClient.post(`/message/forward/draft/${bundleId}/add`, null, {
        params: { messageId }
      });
      return response.data;
    } catch (error) {
      console.error('添加消息到草稿失败:', error);
      throw new Error('添加消息失败');
    }
  }
  
  /**
   * 从草稿移除消息
   */
  async removeMessage(bundleId: string, messageId: number): Promise<{
    success: boolean;
    bundleId: string;
    messageCount: number;
    updatedAt: string;
  }> {
    try {
      const response = await apiClient.post(`/message/forward/draft/${bundleId}/remove`, null, {
        params: { messageId }
      });
      return response.data;
    } catch (error) {
      console.error('从草稿移除消息失败:', error);
      throw new Error('移除消息失败');
    }
  }
  
  /**
   * 更新转发配置
   */
  async updateConfig(bundleId: string, config: ForwardConfig): Promise<{
    success: boolean;
    bundleId: string;
    sendMode: string;
    includeSenderInfo: boolean;
    includeTimestamp: boolean;
    anonymizeSenders: boolean;
  }> {
    try {
      const response = await apiClient.put(`/message/forward/draft/${bundleId}/config`, config);
      return response.data;
    } catch (error) {
      console.error('更新转发配置失败:', error);
      throw new Error('更新配置失败');
    }
  }
  
  /**
   * 执行转发
   */
  async executeForward(
    bundleId: string,
    targetConversationId: number
  ): Promise<{
    success: boolean;
    bundleId: string;
    status: string;
    forwardedAt: string;
    targetConversationId: number;
    error?: string;
  }> {
    try {
      const response = await apiClient.post(`/message/forward/draft/${bundleId}/send`, null, {
        params: { targetConversationId }
      });
      return response.data;
    } catch (error: any) {
      console.error('执行转发失败:', error);
      if (error.response?.data?.error) {
        throw new Error(error.response.data.error);
      }
      throw new Error('执行转发失败');
    }
  }
  
  /**
   * 取消草稿
   */
  async cancelDraft(bundleId: string): Promise<{
    success: boolean;
    bundleId: string;
    cancelledAt: string;
  }> {
    try {
      const response = await apiClient.post(`/message/forward/draft/${bundleId}/cancel`);
      return response.data;
    } catch (error) {
      console.error('取消草稿失败:', error);
      throw new Error('取消草稿失败');
    }
  }
  
  /**
   * 获取草稿列表
   */
  async getDrafts(): Promise<{
    success: boolean;
    count: number;
    drafts: Array<{
      id: number;
      bundleId: string;
      sourceConversationId: number;
      messageCount: number;
      forwardType: string;
      sendMode: string;
      title?: string;
      customComment?: string;
      createdAt: string;
      updatedAt: string;
    }>;
  }> {
    try {
      const response = await apiClient.get('/message/forward/drafts');
      return response.data;
    } catch (error) {
      console.error('获取草稿列表失败:', error);
      throw new Error('获取草稿列表失败');
    }
  }
  
  /**
   * 获取转发历史
   */
  async getHistory(limit: number = 20): Promise<{
    success: boolean;
    count: number;
    history: Array<{
      id: number;
      bundleId: string;
      sourceConversationId: number;
      targetConversationId: number;
      messageCount: number;
      forwardType: string;
      title?: string;
      forwardedAt: string;
    }>;
  }> {
    try {
      const response = await apiClient.get('/message/forward/history', {
        params: { limit }
      });
      return response.data;
    } catch (error) {
      console.error('获取转发历史失败:', error);
      throw new Error('获取转发历史失败');
    }
  }
  
  /**
   * 获取转发统计
   */
  async getStats(): Promise<{
    success: boolean;
    userId: number;
    stats: ForwardStats;
  }> {
    try {
      const response = await apiClient.get('/message/forward/stats');
      return response.data;
    } catch (error) {
      console.error('获取转发统计失败:', error);
      throw new Error('获取转发统计失败');
    }
  }
  
  /**
   * 健康检查
   */
  async healthCheck(): Promise<{
    service: string;
    status: string;
    timestamp: string;
  }> {
    try {
      const response = await apiClient.get('/message/forward/health');
      return response.data;
    } catch (error) {
      console.error('消息转发服务健康检查失败:', error);
      throw new Error('消息转发服务不可用');
    }
  }
  
  /**
   * 生成转发预览
   */
  generateForwardPreview(messageIds: number[], includeSenderInfo: boolean): string {
    // 生成转发预览文本
    const preview = `转发 ${messageIds.length} 条消息${includeSenderInfo ? '（包含发送者信息）' : ''}`;
    return preview;
  }
  
  /**
   * 验证消息选择
   */
  validateMessageSelection(messageIds: number[]): { valid: boolean; error?: string } {
    if (messageIds.length === 0) {
      return { valid: false, error: '请选择至少一条消息' };
    }
    if (messageIds.length > 100) {
      return { valid: false, error: '最多选择 100 条消息' };
    }
    return { valid: true };
  }
}

export const messageForwardService = MessageForwardService.getInstance();

File: im-mobile/lib/services/message_forward_service.dart
/// 消息合并转发服务
/// 支持多条消息合并转发、选择转发、逐条转发

import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

/// 转发包实体
class ForwardBundle {
  final int id;
  final String bundleId;
  final int sourceConversationId;
  final int? targetConversationId;
  final int createdBy;
  final String forwardType;
  final String? title;
  final List<int> messageIds;
  final int messageCount;
  final bool hasMedia;
  final int mediaCount;
  final String status;
  final String sendMode;
  final bool includeSenderInfo;
  final bool includeTimestamp;
  final bool anonymizeSenders;
  final String? customComment;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? forwardedAt;

  ForwardBundle({
    required this.id,
    required this.bundleId,
    required this.sourceConversationId,
    this.targetConversationId,
    required this.createdBy,
    required this.forwardType,
    this.title,
    required this.messageIds,
    required this.messageCount,
    required this.hasMedia,
    required this.mediaCount,
    required this.status,
    required this.sendMode,
    required this.includeSenderInfo,
    required this.includeTimestamp,
    required this.anonymizeSenders,
    this.customComment,
    required this.createdAt,
    required this.updatedAt,
    this.forwardedAt,
  });

  factory ForwardBundle.fromJson(Map<String, dynamic> json) {
    return ForwardBundle(
      id: json['id'] as int,
      bundleId: json['bundleId'] as String,
      sourceConversationId: json['sourceConversationId'] as int,
      targetConversationId: json['targetConversationId'],
      createdBy: json['createdBy'] as int,
      forwardType: json['forwardType'] as String,
      title: json['title'],
      messageIds: (json['messageIds'] as List<dynamic>?)?.map((e) => e as int).toList() ?? [],
      messageCount: json['messageCount'] as int,
      hasMedia: json['hasMedia'] as bool,
      mediaCount: json['mediaCount'] as int,
      status: json['status'] as String,
      sendMode: json['sendMode'] as String,
      includeSenderInfo: json['includeSenderInfo'] as bool,
      includeTimestamp: json['includeTimestamp'] as bool,
      anonymizeSenders: json['anonymizeSenders'] as bool,
      customComment: json['customComment'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      forwardedAt: json['forwardedAt'] != null ? DateTime.parse(json['forwardedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bundleId': bundleId,
      'sourceConversationId': sourceConversationId,
      'targetConversationId': targetConversationId,
      'createdBy': createdBy,
      'forwardType': forwardType,
      'title': title,
      'messageIds': messageIds,
      'messageCount': messageCount,
      'hasMedia': hasMedia,
      'mediaCount': mediaCount,
      'status': status,
      'sendMode': sendMode,
      'includeSenderInfo': includeSenderInfo,
      'includeTimestamp': includeTimestamp,
      'anonymizeSenders': anonymizeSenders,
      'customComment': customComment,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'forwardedAt': forwardedAt?.toIso8601String(),
    };
  }
}

/// 转发配置
class ForwardConfig {
  final String? sendMode;
  final bool? includeSenderInfo;
  final bool? includeTimestamp;
  final bool? anonymizeSenders;
  final String? customComment;
  final String? title;

  ForwardConfig({
    this.sendMode,
    this.includeSenderInfo,
    this.includeTimestamp,
    this.anonymizeSenders,
    this.customComment,
    this.title,
  });

  Map<String, dynamic> toJson() {
    return {
      if (sendMode != null) 'sendMode': sendMode,
      if (includeSenderInfo != null) 'includeSenderInfo': includeSenderInfo,
      if (includeTimestamp != null) 'includeTimestamp': includeTimestamp,
      if (anonymizeSenders != null) 'anonymizeSenders': anonymizeSenders,
      if (customComment != null) 'customComment': customComment,
      if (title != null) 'title': title,
    };
  }
}

/// 消息合并转发服务
class MessageForwardService {
  static final MessageForwardService _instance = MessageForwardService._internal();
  
  factory MessageForwardService() {
    return _instance;
  }
  
  MessageForwardService._internal();
  
  final String _baseUrl = 'https://api.example.com/api/message/forward';
  late String _authToken;
  
  /// 初始化服务
  void initialize(String authToken) {
    _authToken = authToken;
  }
  
  /// 创建转发草稿
  Future<Map<String, dynamic>> createDraft({
    required int sourceConversationId,
    required String forwardType,
    required List<int> messageIds,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/draft').replace(queryParameters: {
          'sourceConversationId': sourceConversationId.toString(),
          'forwardType': forwardType,
        }),
        headers: _buildHeaders(),
        body: jsonEncode(messageIds),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('创建转发草稿失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('创建转发草稿失败：$e');
      rethrow;
    }
  }
  
  /// 添加消息到草稿
  Future<Map<String, dynamic>> addMessage(String bundleId, int messageId) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/draft/$bundleId/add').replace(queryParameters: {
          'messageId': messageId.toString(),
        }),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('添加消息失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('添加消息失败：$e');
      rethrow;
    }
  }
  
  /// 从草稿移除消息
  Future<Map<String, dynamic>> removeMessage(String bundleId, int messageId) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/draft/$bundleId/remove').replace(queryParameters: {
          'messageId': messageId.toString(),
        }),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('移除消息失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('移除消息失败：$e');
      rethrow;
    }
  }
  
  /// 更新转发配置
  Future<Map<String, dynamic>> updateConfig(String bundleId, ForwardConfig config) async {
    try {
      final response = await http.put(
        Uri.parse('$_baseUrl/draft/$bundleId/config'),
        headers: _buildHeaders(),
        body: jsonEncode(config.toJson()),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('更新配置失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('更新配置失败：$e');
      rethrow;
    }
  }
  
  /// 执行转发
  Future<Map<String, dynamic>> executeForward(String bundleId, int targetConversationId) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/draft/$bundleId/send').replace(queryParameters: {
          'targetConversationId': targetConversationId.toString(),
        }),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        final errorData = jsonDecode(response.body);
        throw Exception(errorData['error'] ?? '执行转发失败');
      }
    } catch (e) {
      debugPrint('执行转发失败：$e');
      rethrow;
    }
  }
  
  /// 取消草稿
  Future<Map<String, dynamic>> cancelDraft(String bundleId) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/draft/$bundleId/cancel'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('取消草稿失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('取消草稿失败：$e');
      rethrow;
    }
  }
  
  /// 获取草稿列表
  Future<Map<String, dynamic>> getDrafts() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/drafts'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取草稿列表失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取草稿列表失败：$e');
      rethrow;
    }
  }
  
  /// 获取转发历史
  Future<Map<String, dynamic>> getHistory({int limit = 20}) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/history').replace(queryParameters: {
          'limit': limit.toString(),
        }),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取转发历史失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取转发历史失败：$e');
      rethrow;
    }
  }
  
  /// 获取转发统计
  Future<Map<String, dynamic>> getStats() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/stats'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取转发统计失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取转发统计失败：$e');
      rethrow;
    }
  }
  
  /// 健康检查
  Future<Map<String, dynamic>> healthCheck() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/health'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('健康检查失败：${response.statusCode}');
      }
    } catch (e) {
      debugPrint('健康检查失败：$e');
      rethrow;
    }
  }
  
  /// 验证消息选择
  Map<String, dynamic> validateMessageSelection(List<int> messageIds) {
    if (messageIds.isEmpty) {
      return {'valid': false, 'error': '请选择至少一条消息'};
    }
    if (messageIds.length > 100) {
      return {'valid': false, 'error': '最多选择 100 条消息'};
    }
    return {'valid': true};
  }
  
  /// 构建请求头
  Map<String, String> _buildHeaders() {
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $_authToken',
      'User-Agent': 'IM-Mobile-App',
    };
  }
}

/// 全局服务实例
final messageForwardService = MessageForwardService();